File: src/main/java/com/example/chatService/kafka/DMOutboxProcessor.java
package com.example.chatService.kafka;

import com.example.chatService.dto.DMMessageKafkaDto;
import com.example.chatService.dto.MessagingStatus;
import com.example.chatService.entity.DMOutbox;
import com.example.chatService.repository.DMOutboxRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Slf4j
@RequiredArgsConstructor
public class DMOutboxProcessor {

    private final DMOutboxRepository outboxRepository;
    private final KafkaTemplate<String, DMMessageKafkaDto> kafkaTemplate;

    private static final String TOPIC = "dm-messages";
    private static final int BATCH_SIZE = 100;

    @Value("${app.instance-id")
    private String workerId;

    @Transactional
    @Scheduled(fixedDelay = 50)
    public void processOutbox() throws Exception {

        int claimed = outboxRepository.claimBatch(workerId, BATCH_SIZE);
        if (claimed == 0) return;

        List<DMOutbox> list = outboxRepository
                .findByStatusAndLockedByOrderByIdAsc(MessagingStatus.PROCESSING, workerId);

        for (DMOutbox box : list) {
            try {
                DMMessageKafkaDto message = DMMessageKafkaDto.builder()
                        .roomId(box.getRoomId())
                        .senderId(box.getSenderId())
                        .content(box.getContent())
                        .sentAt(box.getEventTimestamp())
                        .build();

                kafkaTemplate.send(TOPIC, box.getRoomId(), message).get();

                box.setStatus(MessagingStatus.SENT);
                box.setLockedBy(null);
                box.setLockedAt(null);

            } catch (Exception e) {
                // 실패시 다시 NEW
                box.setStatus(MessagingStatus.NEW);
                box.setLockedBy(null);
                box.setLockedAt(null);
            }
        }
    }
}

File: src/main/java/com/example/chatService/kafka/DMOutboxProcessor.kt
package com.example.chatService.kafka

import com.example.chatService.dto.DMMessageKafkaDto
import com.example.chatService.dto.MessagingStatus
import com.example.chatService.entity.DMOutbox
import com.example.chatService.repository.DMOutboxRepository
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class DMOutboxProcessor(
        private val outboxRepository: DMOutboxRepository,
        private val kafkaTemplate: KafkaTemplate<String, DMMessageKafkaDto>
) {

    companion object {
        private const val TOPIC = "dm-messages"
        private const val BATCH_SIZE = 100
    }

    @Value("\${app.instance-id}")
    private lateinit var workerId: String

    @Transactional
    @Scheduled(fixedDelay = 50)
    @Throws(Exception::class)
    fun processOutbox() {
        val claimed = outboxRepository.claimBatch(workerId, BATCH_SIZE)
        if (claimed == 0) return

        val list: List<DMOutbox> =
                outboxRepository.findByStatusAndLockedByOrderByIdAsc(
                        MessagingStatus.PROCESSING,
                        workerId
                )

        for (box in list) {
            try {

                val roomId = box.roomId ?: throw IllegalStateException("roomId is null")
                val senderId = box.senderId ?: throw IllegalStateException("senderId is null")
                val content = box.content ?: throw IllegalStateException("content is null")

                val message = DMMessageKafkaDto(
                        roomId = roomId,
                        senderId = senderId,
                        content = content,
                        sentAt = box.eventTimestamp
                )

                kafkaTemplate.send(TOPIC, box.roomId, message).get()

                box.status = MessagingStatus.SENT
                box.lockedBy = null
                box.lockedAt = null

            } catch (e: Exception) {
                box.status = MessagingStatus.NEW
                box.lockedBy = null
                box.lockedAt = null
            }
        }
    }
}